indra/llprimitive/llprimitive.cpp

```diff
@@ -159,6 +162,71 @@ const U8 material_id_type::s_null_id[] = { 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 };
 // TODO -- eliminate this global from the codebase!
 LLVolumeMgr* LLPrimitive::sVolumeManager = NULL;
 
+LLTEContents::LLTEContents(size_t N)
+{
+    llassert(N > 0);
+    N = std::min(N, size_t(LLTEContents::MAX_TES));
+    num_textures = N;
+
+    // allocate one big buffer for all the data and fill it with zeros
+    size_t bytes_per_texture = sizeof(LLUUID)
+        + sizeof(LLMaterialID)
+        + sizeof(LLColor4U)
+        + 2 * sizeof(F32)
+        + 3 * sizeof(S16)
+        + 4 * sizeof(U8);
+    size_t num_bytes = num_textures * bytes_per_texture;
+    data = new U8[num_bytes];
+    memset(data, 0, num_bytes);
+
+    // compute offset pointers into data for the various fields
+    size_t offset = 0;
+    image_ids = reinterpret_cast<LLUUID*>(data);
+    offset += num_textures * sizeof(LLUUID);
+
+    material_ids = reinterpret_cast<LLMaterialID*>(data + offset);
+    offset += num_textures * sizeof(LLMaterialID);
+
+    colors = reinterpret_cast<LLColor4U*>(data + offset);
+    offset += num_textures * sizeof(LLColor4U);
+
+    scale_s = reinterpret_cast<F32*>(data + offset);
+    offset += num_textures * sizeof(F32);
+
+    scale_t = reinterpret_cast<F32*>(data + offset);
+    offset += num_textures * sizeof(F32);
+
+    offset_s = reinterpret_cast<S16*>(data + offset);
+    offset += num_textures * sizeof(S16);
+
+    offset_t = reinterpret_cast<S16*>(data + offset);
+    offset += num_textures * sizeof(S16);
+
+    rot = reinterpret_cast<S16*>(data + offset);
+    offset += num_textures * sizeof(S16);
+
+    bump = reinterpret_cast<U8*>(data + offset);
+    offset += num_textures * sizeof(U8);
+
+    media_flags = reinterpret_cast<U8*>(data + offset);
+    offset += num_textures * sizeof(U8);
+
+    glow = reinterpret_cast<U8*>(data + offset);
+    offset += num_textures * sizeof(U8);
+
+    alpha_gamma = reinterpret_cast<U8*>(data + offset);
+    offset += num_textures * sizeof(U8);
+
+    // verify we correctly computed data size
+    llassert(offset == num_bytes);
+}
```

Comment:
`LLTEContents` ctor now takes an argument for the number of TEs to expect.  This allows it to only allocate as much memory as it actually needs (e.g. 6 for a box) instead of the default (45, which is how many the avatar uses).